src/go/wsl-helper/pkg/wsl-utils/version.go

```diff
@@ -0,0 +1,292 @@
+		// This is expected: we didn't provide any buffer
+		break
+	default:
+		return nil, fmt.Errorf("error getting buffer size: %w: %w", windows.Errno(rv), err)
+	}
+
+	packageNames := make([]uintptr, count)
+	packageNameBuffer := make([]uint16, bufferLength)
+
+	rv, _, err = getPackagesByPackageFamily.Call(
+		uintptr(unsafe.Pointer(packageFamilyBytes)),
+		uintptr(unsafe.Pointer(&count)),
+		uintptr(unsafe.Pointer(unsafe.SliceData(packageNames))),
+		uintptr(unsafe.Pointer(&bufferLength)),
+		uintptr(unsafe.Pointer(unsafe.SliceData(packageNameBuffer))),
+	)
+	if rv != uintptr(windows.ERROR_SUCCESS) {
+		return nil, fmt.Errorf("error getting package names: %w: %w", windows.Errno(rv), err)
+	}
+
+	result := make([]string, count)
+	slice := unsafe.Slice((**uint16)(unsafe.Pointer(unsafe.SliceData(packageNames))), count)
+	for i, ptr := range slice {
+		result[i] = windows.UTF16PtrToString(ptr)
+	}
+
+	return result, nil
+}
+
+// PackageVersion corresponds to the PACKAGE_VERSION structure.
+type PackageVersion struct {
+	Revision uint16 `json:"revision"`
+	Build    uint16 `json:"build"`
+	Minor    uint16 `json:"minor"`
+	Major    uint16 `json:"major"`
+}
+
+func (v PackageVersion) String() string {
+	return fmt.Sprintf("%d.%d.%d.%d", v.Major, v.Minor, v.Build, v.Revision)
+}
+
+// packageInfo corresponds to the PACKAGE_INFO structure.
+type packageInfo struct {
+	reserved          uint32
+	flags             uint32
+	path              *uint16
+	packageFullName   *uint16
+	packageFamilyName *uint16
+	packageId         struct {
+		reserved              uint32
+		processorArchitecture uint32
+		version               PackageVersion
+		name                  *uint16
+		publisher             *uint16
+		resourceId            *uint16
+		publisherId           *uint16
+	}
+}
+
+// getPackageVersion gets the package version of the package with the given
+// full name.
+func getPackageVersion(fullName string) (*PackageVersion, error) {
+	nameBuffer, err := windows.UTF16PtrFromString(fullName)
+	if err != nil {
+		return nil, err
+	}
+	var packageInfoReference uintptr
+	rv, _, err := openPackageInfoByFullName.Call(
+		uintptr(unsafe.Pointer(nameBuffer)),
+		0, // reserved
+		uintptr(unsafe.Pointer(&packageInfoReference)),
+	)
+	if rv != uintptr(windows.ERROR_SUCCESS) {
+		return nil, fmt.Errorf("error opening package info: %w: %w", windows.Errno(rv), err)
+	}
+	defer closePackageInfo.Call(packageInfoReference)
+
+	var bufferLength, count uint32
+	rv, _, err = getPackageInfo.Call(
+		packageInfoReference,
+		uintptr(PACKAGE_INFORMATION_BASIC),
+		uintptr(unsafe.Pointer(&bufferLength)),
+		uintptr(unsafe.Pointer(nil)),
+		uintptr(unsafe.Pointer(nil)),
+	)
+	switch rv {
+	case uintptr(windows.ERROR_SUCCESS):
+		break
+	case uintptr(windows.ERROR_INSUFFICIENT_BUFFER):
+		// This is expected: we didn't provide any buffer
+		break
+	default:
+		return nil, fmt.Errorf("error getting buffer size: %w: %w", windows.Errno(rv), err)
```

Comment:
I feel like this error could be declared into an error type or a var that could get re-used.

Reply:
Sorry, could you clarify? `windows.Errno(rv)` can't be extracted out because `rv` was assigned to a couple line above, and so was `err`; the `fmt.Errorf` string is unique (within this function at least); I guess I can have a function that takes `rv` and `err`, plus a string, and return the result?